File: server/modules/terminal/ring-buffer.ts
export class RingBuffer {
  private chunks: Buffer[] = [];
  private totalBytes = 0;

  constructor(private readonly capBytes: number) {}

  append(data: string | Buffer): void {
    const buf = typeof data === "string" ? Buffer.from(data, "utf-8") : data;
    if (buf.length === 0) return;
    this.chunks.push(buf);
    this.totalBytes += buf.length;
    while (this.totalBytes > this.capBytes && this.chunks.length > 0) {
      const removed = this.chunks.shift()!;
      this.totalBytes -= removed.length;
    }
  }

  snapshot(): Buffer {
    if (this.chunks.length === 0) return Buffer.alloc(0);
    return Buffer.concat(this.chunks);
  }

  get byteLength(): number {
    return this.totalBytes;
  }
}

File: server-rs/src/terminal/ring_buffer.rs
/// Byte-capped ring buffer for terminal scrollback (matches server ring-buffer.ts).
pub struct RingBuffer {
    chunks: Vec<Vec<u8>>,
    total_bytes: usize,
    cap_bytes: usize,
}

impl RingBuffer {
    pub fn new(cap_bytes: usize) -> Self {
        Self {
            chunks: Vec::new(),
            total_bytes: 0,
            cap_bytes,
        }
    }

    pub fn append(&mut self, data: impl AsRef<[u8]>) {
        let buf = data.as_ref();
        if buf.is_empty() {
            return;
        }
        self.chunks.push(buf.to_vec());
        self.total_bytes += buf.len();
        while self.total_bytes > self.cap_bytes && !self.chunks.is_empty() {
            let removed = self.chunks.remove(0);
            self.total_bytes -= removed.len();
        }
    }

    pub fn snapshot(&self) -> Vec<u8> {
        if self.chunks.is_empty() {
            return Vec::new();
        }
        self.chunks.iter().flatten().copied().collect()
    }

    pub fn byte_length(&self) -> usize {
        self.total_bytes
    }
}
